### PR TITLE
BulkUpdateRecordStatusForPublisFlag

### PR DESCRIPTION
**Pull request by MTSarwar**

### CODE REVIEW

File: src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs [hunk 1]
```diff
@@ -386,5 +391,68 @@ public async Task<int> GetTotalLiveApprenticeships(DocumentClient client, string
...
+        {       
+            string scriptFileName = @"Data/StoreProcedureApprenticeship_ChangeRecordStatus.js";
+            string StoredProcedureName = Path.GetFileNameWithoutExtension(scriptFileName);          
+
+            await CreateStoredProcedureIfNotExistsAsync(GetClient(), _settings.DatabaseId, StoredProcedureName, scriptFileName);
+        }
+      
+        public async Task CreateStoredProcedureIfNotExistsAsync(DocumentClient client, string collectionId, string procedureName, string procedurePath)
+        {
+            var errorMessage = String.Empty;
```
Comment (gunndabad) [inline src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs]: Don't need these two lines

File: src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs [hunk 2]
```diff
@@ -386,5 +391,68 @@ public async Task<int> GetTotalLiveApprenticeships(DocumentClient client, string
...
+            StoredProcedure isStoredProcedureExist = client.CreateStoredProcedureQuery(collectionLink)
+                                   .Where(sp => sp.Id == StoredProcedureName)
+                                   .AsEnumerable()
+                                   .FirstOrDefault();                        
+            try
+            {
+                if (isStoredProcedureExist == null)
+                {
+                    Assembly assembly = this.GetType().Assembly;
+                    var resourceStream = assembly.GetManifestResourceStream(assembly.GetName().Name + "." + "Data.StoreProcedure" + ".Apprenticeship_ChangeRecordStatus.js");
```
Comment (gunndabad) [inline src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs]: Need a `using` statement around this stream

File: src/Dfc.ProviderPortal.Apprenticeships/Data/StoreProcedure/Apprenticeship_ChangeRecordStatus.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                for (var i = 0; i < documents.length; i++)
+                    tryUpdate(documents[i]);
+                tryQueryAndUpdate(responseOptions.continuation);
+            } else if (responseOptions.continuation) {
+                // Else if the query came back empty, but with a continuation token; repeat the query w/ the token.
+                // It is highly unlikely for this to happen when performing a query by id; but is included to serve as an example for larger queries.
+                responseBody.log += "Continue query";
+                tryQueryAndUpdate(responseOptions.continuation);
+            } else {
+                // responseBody.log += "No documents found";
```
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Data/StoreProcedure/Apprenticeship_ChangeRecordStatus.js]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Data/StoreProcedure/Apprenticeship_ChangeRecordStatus.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
+﻿
+function updateSproc(UKPRN, currentStatus, statusToBeChangedTo) {
+    var collection = getContext().getCollection();
+    var collectionLink = collection.getSelfLink();
+    var response = getContext().getResponse();
+
+    var responseBody = {
```
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Data/StoreProcedure/Apprenticeship_ChangeRecordStatus.js]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Data/StoreProcedure/Apprenticeship_ChangeRecordStatus.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                // Else if the query came back empty, but with a continuation token; repeat the query w/ the token.
+                // It is highly unlikely for this to happen when performing a query by id; but is included to serve as an example for larger queries.
+                responseBody.log += "Continue query";
+                tryQueryAndUpdate(responseOptions.continuation);
+            } else {
+                // responseBody.log += "No documents found";
+                //responseBody.continuation = false;
+                // response.setBody(responseBody);
+                // response.setBody({ updated: updated });
+                response.setBody(updated)
```
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Data/StoreProcedure/Apprenticeship_ChangeRecordStatus.js]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Dfc.ProviderPortal.Apprenticeships.csproj [hunk 6]
```diff
@@ -5,11 +5,19 @@
     <AssemblyVersion>1.0.19189.1347</AssemblyVersion>
     <FileVersion>1.0.19189.1347</FileVersion>
   </PropertyGroup>
+  <ItemGroup>
+    <None Remove="Data\StoreProcedure\Apprenticeship_ChangeRecordStatus.js" />
```
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Dfc.ProviderPortal.Apprenticeships.csproj]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Services/ApprenticeshipService.cs [hunk 7]
```diff
@@ -247,6 +247,30 @@ public async Task<IApprenticeship> Update(IApprenticeship apprenticeship)
 
         }
 
+        //public async Task<HttpResponseMessage> ChangeApprenticeshipStatusForUKPRNSelection(int UKPRN, RecordStatus CurrentStatus, RecordStatus StatusToBeChangedTo)
```
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Services/ApprenticeshipService.cs]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/WebJobsExtensionStartup.cs [hunk 8]
```diff
@@ -11,13 +11,15 @@
...
 using Dfc.ProviderPortal.Apprenticeships.Interfaces.Apprenticeships;
+using Dfc.ProviderPortal.Apprenticeships.Interfaces.Settings;
 
 [assembly: WebJobsStartup(typeof(WebJobsExtensionStartup), "Web Jobs Extension Startup")]
 
 namespace Dfc.ProviderPortal.Apprenticeships
 {
     public class WebJobsExtensionStartup : IWebJobsStartup
     {
+        private readonly ICosmosDbSettings _settings;
```
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/WebJobsExtensionStartup.cs]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Interfaces/Helper/ICosmosDbHelper.cs [hunk 9]
```diff
@@ -15,6 +17,7 @@ public interface ICosmosDbHelper
         Task<Database> CreateDatabaseIfNotExistsAsync(DocumentClient client);
         Task<DocumentCollection> CreateDocumentCollectionIfNotExistsAsync(DocumentClient client, string collectionId);
         Task<Document> CreateDocumentAsync(DocumentClient client, string collectionId, object document);
+        Task CreateStoredProcedureIfNotExistsAsync(DocumentClient client, string collectionId, string procedureName, string procedurePath);
```
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Interfaces/Helper/ICosmosDbHelper.cs]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Interfaces/Helper/ICosmosDbHelper.cs [hunk 10]
```diff
@@ -31,5 +34,9 @@ public interface ICosmosDbHelper
         Task<List<ApprenticeshipDfcReportDocument>> GetAllDfcMigrationReports(DocumentClient client,
             string collectionId);
         Task<int> GetTotalLiveApprenticeships(DocumentClient client, string collectionId);
+        Task<int> ExecuteStoredProcedureAsync(DocumentClient client, string collectionId, string procedureName, int UKPRN, int currentStatus, int statusToBeChangedTo, int partitionKey);
+
+        Task Initialise_CreateStoreProc();
```
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Interfaces/Helper/ICosmosDbHelper.cs]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Interfaces/Helper/ICosmosDbHelper.cs [hunk 11]
```diff
@@ -31,5 +34,9 @@ public interface ICosmosDbHelper
         Task<List<ApprenticeshipDfcReportDocument>> GetAllDfcMigrationReports(DocumentClient client,
             string collectionId);
         Task<int> GetTotalLiveApprenticeships(DocumentClient client, string collectionId);
+        Task<int> ExecuteStoredProcedureAsync(DocumentClient client, string collectionId, string procedureName, int UKPRN, int currentStatus, int statusToBeChangedTo, int partitionKey);
```
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Interfaces/Helper/ICosmosDbHelper.cs]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs [hunk 12]
```diff
@@ -386,5 +391,68 @@ public async Task<int> GetTotalLiveApprenticeships(DocumentClient client, string
...
+                    }
+
+                    StoredProcedure sproc = await client.CreateStoredProcedureAsync(collectionLink, new StoredProcedure
+                    {
+                        Id = StoredProcedureName,
+                        Body = sProcresult
+                    });
+                }
+            }
+            catch(Exception ex)
```
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs [hunk 13]
```diff
@@ -386,5 +391,68 @@ public async Task<int> GetTotalLiveApprenticeships(DocumentClient client, string
...
+      
+        public async Task CreateStoredProcedureIfNotExistsAsync(DocumentClient client, string collectionId, string procedureName, string procedurePath)
+        {
+            var errorMessage = String.Empty;
+            string sProcresult;
+            Throw.IfNull(client, nameof(client));
+            Throw.IfNullOrWhiteSpace(collectionId, nameof(collectionId));
+
+            string scriptFileName = @"/Data/StoreProcedure/Apprenticeship_ChangeRecordStatus";
+            string StoredProcedureName = Path.GetFileNameWithoutExtension(scriptFileName);
```
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs [hunk 14]
```diff
@@ -386,5 +391,68 @@ public async Task<int> GetTotalLiveApprenticeships(DocumentClient client, string
...
+        }
+      
+        public async Task CreateStoredProcedureIfNotExistsAsync(DocumentClient client, string collectionId, string procedureName, string procedurePath)
+        {
+            var errorMessage = String.Empty;
+            string sProcresult;
+            Throw.IfNull(client, nameof(client));
+            Throw.IfNullOrWhiteSpace(collectionId, nameof(collectionId));
+
+            string scriptFileName = @"/Data/StoreProcedure/Apprenticeship_ChangeRecordStatus";
```
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs (hunk 1 above)
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs (hunk 2 above)
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Data/StoreProcedure/Apprenticeship_ChangeRecordStatus.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Data/StoreProcedure/Apprenticeship_ChangeRecordStatus.js]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Data/StoreProcedure/Apprenticeship_ChangeRecordStatus.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Data/StoreProcedure/Apprenticeship_ChangeRecordStatus.js]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Data/StoreProcedure/Apprenticeship_ChangeRecordStatus.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Data/StoreProcedure/Apprenticeship_ChangeRecordStatus.js]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Dfc.ProviderPortal.Apprenticeships.csproj (hunk 6 above)
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Dfc.ProviderPortal.Apprenticeships.csproj]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs]: Done

File: src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs (hunk 2 above)
Comment (MTSarwar) [reply src/Dfc.ProviderPortal.Apprenticeships/Helper/CosmosDbHelper.cs]: Done

**Pull request merged**